hi-ml-histopathology/src/histopathology/utils/output_utils.py

```diff
@@ -0,0 +1,385 @@
+
+
+def save_confusion_matrix(conf_matrix_metric: ConfusionMatrix, class_names: Sequence[str], figures_dir: Path) -> None:
+    print("Computing and saving confusion matrix...")
+    cf_matrix = conf_matrix_metric.compute().cpu().numpy()
+    #  We can't log tensors in the normal way - just print it to console
+    print('test/confusion matrix:')
+    print(cf_matrix)
+    #  Save the normalized confusion matrix as a figure in outputs
+    cf_matrix_n = cf_matrix / cf_matrix.sum(axis=1, keepdims=True)
+    fig = plot_normalized_confusion_matrix(cm=cf_matrix_n, class_names=(class_names))
+    save_figure(fig=fig, figpath=figures_dir / 'normalized_confusion_matrix.png')
+
+
+class OutputsPolicy:
+    """Utility class that defines when to save validation epoch outputs."""
+
+    _BEST_EPOCH_KEY = 'best_epoch'
+    _BEST_VALUE_KEY = 'best_value'
+    _PRIMARY_METRIC_KEY = 'primary_metric'
+
+    def __init__(self, outputs_root: Path, primary_val_metric: MetricsKey, maximise: bool) -> None:
+        """
+        :param outputs_root: Root directory where to save a recovery file with best epoch and metric value.
+        :param primary_val_metric: Name of the validation metric to track for saving best epoch outputs.
+        :param maximise: Whether higher is better for `primary_val_metric`.
+        """
+        self.outputs_root = outputs_root
+        self.primary_val_metric = primary_val_metric
+        self.maximise = maximise
+
+        self._init_best_metric()
+
+    @property
+    def best_metric_file_path(self) -> Path:
+        return self.outputs_root / "best_val_metric.yml"
+
+    def _init_best_metric(self) -> None:
+        """Initialise running best metric epoch and value (recovered from disk if available).
+
+        :raises ValueError: If the primary metric name does not match the one saved on disk.
+        """
+        if self.best_metric_file_path.exists():
+            contents = YAML().load(self.best_metric_file_path)
+            self._best_metric_epoch = contents[self._BEST_EPOCH_KEY]
+            self._best_metric_value = contents[self._BEST_VALUE_KEY]
+            if contents[self._PRIMARY_METRIC_KEY] != self.primary_val_metric:
+                raise ValueError(f"Expected primary metric '{self.primary_val_metric}', but found "
+                                 f"'{contents[self._PRIMARY_METRIC_KEY]}' in {self.best_metric_file_path}")
+        else:
+            self._best_metric_epoch = 0
+            self._best_metric_value = float('-inf') if self.maximise else float('inf')
+
+    def _save_best_metric(self) -> None:
+        """Save best metric epoch, value, and name to disk, to allow recovery (e.g. in case of pre-emption)."""
+        contents = {self._BEST_EPOCH_KEY: self._best_metric_epoch,
+                    self._BEST_VALUE_KEY: self._best_metric_value,
+                    self._PRIMARY_METRIC_KEY: self.primary_val_metric.value}
+        YAML().dump(contents, self.best_metric_file_path)
```

Comment:
why do we need this YAML() here? At other places in our code, we just use ruamel.yaml.dump?

Reply:
I was following the recommendations from the `ruamel` docs: https://yaml.readthedocs.io/en/latest/api.html